File: 0075. Sort Colors.js
// INPUT: [ 2, 0, 2, 1, 1, 0 ]
// OUTPUT: [ 0, 0, 1, 1, 2, 2 ]

const sortColors = (nums) => {
  let idx = (start = 0),
    end = nums.length - 1;

  while (idx <= end) {
    if (nums[idx] === 0) {
      nums[idx] = 1;
      nums[start] = 0;
      start++;
      idx++;
    } else if (nums[idx] === 2) {
      nums[idx] = nums[end];
      nums[end] = 2;
      end--;
    } else idx++;
  }

  return nums;
};

console.log(sortColors([2, 0, 2, 1, 1, 0]));

File: python/0075. Sort Colors.py
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        start = 0
        mid = 0
        end = len(nums) - 1

        while mid <= end:
            if nums[mid] == 0:
                nums[start], nums[mid] = nums[mid], nums[start]
                start += 1
                mid += 1
            elif nums[mid] == 2:
                nums[mid], nums[end] = nums[end], nums[mid]
                end -= 1
            else:
                mid += 1
